### PR TITLE
update checkpoint replay status

### PR DESCRIPTION
**Pull request by ide-rea**
As checkpoint is a part of wal replay, so also include checkpoint replay status.

### CODE REVIEW

**Comment by roidelapluie** [comment]
cc @bwplotka @codesome can you please look at this?

File: tsdb/head.go [hunk 1]
```diff
@@ -805,6 +805,26 @@ func (h *Head) Init(minValidTime int64) error {
...
+	// Find the last segment.
+	// The purpose of this scope is to avoid wal.Segments err override wal.LastCheckpoint err.
+	var endAt int
+	{
+		var err error
+		_, endAt, err = wal.Segments(h.wal.Dir())
+		if err != nil {
+			return errors.Wrap(err, "finding WAL segments")
+		}
+	}
```
Comment (codesome) [inline tsdb/head.go:818]: Not so comfortable having `err` in between when we are consuming some other `err` below this. How about this?
```suggestion
	_, endAt, e := wal.Segments(h.wal.Dir())
	if e != nil {
		return errors.Wrap(e, "finding WAL segments")
	}
```

File: tsdb/head.go [hunk 2]
```diff
@@ -805,6 +805,26 @@ func (h *Head) Init(minValidTime int64) error {
...
+
+	// Not likely endAt will smaller then startFrom,
+	// as the head cut a new wal segment file before creating a new checkpoint.
+	if endAt < startFrom {
+		endAt = startFrom
+	}
+
+	h.startWALReplayStatus(startFrom, endAt)
```
Comment (codesome) [inline tsdb/head.go:826]: I am not sure where is it actually considering the checkpoint. From what I understand, the status starts from `startFrom`, and even after checkpoint replay it is still `startFrom`. Only thing that changes is the startup screen comes up early?

File: tsdb/head.go [hunk 3]
```diff
@@ -805,6 +805,26 @@ func (h *Head) Init(minValidTime int64) error {
...
+		if err != nil {
+			return errors.Wrap(err, "finding WAL segments")
+		}
+	}
+
+	// Not likely endAt will smaller then startFrom,
+	// as the head cut a new wal segment file before creating a new checkpoint.
+	if endAt < startFrom {
+		endAt = startFrom
+	}
```
Comment (codesome) [inline tsdb/head.go:824]: If someone manually deletes some WAL files (which is not so rare), this case is possible and it changes the behaviour below and can cause errors. We should not change this here, and instead handle the min being greater than max in `startWALReplayStatus`.

**Review by codesome — COMMENTED** [review]
Thanks for the PR. I have a question below on how exactly is it considering the checkpoint here.

File: tsdb/head.go (hunk 2 above)
Comment (ide-rea) [reply tsdb/head.go:822]: > From what I understand, the status starts from `startFrom`, and even after checkpoint replay it is still `startFrom`
What i understand is the `startFrom` will be `startFrom++` after the checkpoint replay. 
That's right At the front end, 

> the startup screen comes up early?

File: tsdb/head.go (hunk 3 above)
Comment (ide-rea) [reply tsdb/head.go:820]: Very nice consideration 👍

**Review by codesome — APPROVED** [review]
LGTM, Thanks! Can you please [fix the DCO](https://github.com/prometheus/prometheus/pull/8898/checks?check_run_id=3034384573)?

**Comment by ide-rea** [comment]
Very thanks for the review! But the circleci test not passed which may related to [this known issue](https://github.com/prometheus/prometheus/issues/7672)

**Comment by ide-rea** [comment]
Very thanks for the review! 
But, the circleci test not passed which seems related to this issue #7672

**Pull request merged**